### PR TITLE
optimizer: make cardinality estimates fully concrete

### PR DESCRIPTION
**Pull request by mgree**
Cardinality estimation uses symbolic arithmetic expressions when computing cardinality, which can lead to some enormous expressions when working with disjunctions.

Postgres just works with concrete floating point numbers as selectivity ([`typedef`](https://github.com/postgres/postgres/blob/fbff304c574971d15981a080cea8d5a710b81d4d/src/include/nodes/nodes.h#L250)) and will [calculate selectivity for disjunctions as though they were independent](https://github.com/postgres/postgres/blob/fbff304c574971d15981a080cea8d5a710b81d4d/src/backend/optimizer/path/clausesel.c#L415).

We can avoid the memory blow-up by being concrete throughout, like postgres.

There's a subtle change in behavior with this refactor: if the system variable `enable_cardinality_estimates` is set, some cardinalities may be reported even though the session variable `enable_session_cardinality_estimates` is not set. The refactored `Optimizer` interface takes its defaults from system variables (and a few other places), but not session variables. Ordinarily I'd strive to match the original behavior letter-for-letter, but the current ethic of "session variables don't affect the optimizer" seems like a prudent one.

### Motivation

  * This PR fixes a recognized bug.
    + #25805
    + #25769

### Tips for reviewer

I am sorry that I didn't break this up into more sensible diffs. There are a few classes of change:

  - adapter changes to plumb stats properly
  - re-enabling things in `JoinImplementation`
  - simplification to the `Cardinality` analysis

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/transform/src/join_implementation.rs [hunk 1]
```diff
@@ -268,10 +270,24 @@ impl JoinImplementation {
                 }
                 let push_down_characteristics =
                     FilterCharacteristics::filter_characteristics(&push_downs[index])?;
-                // let push_down_factor = push_down_characteristics.worst_case_scaling_factor();
```
Comment (ggevay) [inline src/transform/src/join_implementation.rs:271]: Is there a reason why `push_down_characteristics` is not guarded by the below `if features.enable_cardinality_estimates`?

File: src/transform/src/analysis.rs [hunk 2]
```diff
@@ -1213,125 +1213,209 @@ mod cardinality {
...
+
+    impl From<f64> for CardinalityEstimate {
+        fn from(value: f64) -> Self {
+            Self::Estimate(OrderedFloat(value))
+        }
     }
 
     /// The default selectivity for predicates we know nothing about.
     ///
     /// It is safe to use this instead of `FactorizerVariable::Index(col)`.
```
Comment (ggevay) [inline src/transform/src/analysis.rs:1389]: Stale comment?

File: src/transform/src/analysis.rs [hunk 3]
```diff
@@ -1340,61 +1424,53 @@ mod cardinality {
...
-                        }
-
-                        factor
-                    }
+                    VariadicFunc::And => exprs
+                        .iter()
+                        .map(|expr| self.predicate(expr, unique_columns))
+                        .product(),
                     VariadicFunc::Or => {
                         // TODO(mgree): BETWEEN will get compiled down to an OR of appropriate bounds---we could try to detect it and be clever
```
Comment (ggevay) [inline src/transform/src/analysis.rs:1476]: > OR of appropriate bounds

AND

**Review by ParkMyCar — APPROVED** [review]
Only reviewed the Adapter changes, they LGTM

**Comment by mgree** [comment]
> I've kicked off a Nightly, and a full slt:
> 
> * https://buildkite.com/materialize/nightly/builds/8046
> * https://buildkite.com/materialize/slt/builds/5952
> 
> The failure in the full slt run in "SQL logic tests 6" should go away after a rebase on main.
> 
> Edit: There seem to be also lots of Nightly failures. Let's run Nightly again after a rebase on main.

Will do!

> If I'm reading the code correctly, the enabling of cardinality estimation in the join planning still wouldn't affect production workloads, because that code is guarded by `if features.enable_cardinality_estimates`. So, `enable_cardinality_estimates` must have been On in the tests where #25769 was happening, right? This is because, if I remember correctly, all feature flags are on in most of our CI, right?

Yes, I believe that's correct. I've tested locally with and without the flag and see very fast response times.

> Another thing: [Here](https://github.com/MaterializeInc/materialize/issues/25805#issuecomment-1988506073), you wrote that even with concrete cardinality estimation, we would still have a combinatorial amount of work to do. How did you solve this problem?

I was wrong about it being a combinatorial amount of work: when we work concretely, we do linear work in the number of disjunctions---we're just keeping floats around. The old version doubled the memory every time when it cloned `expr1` and did linear work in the size of `expr1` for the addition and multiplication---this quickly became exponential. But we're good now: it's constant work per disjunction/iteration, and linear overall.

File: src/transform/src/join_implementation.rs (hunk 1 above)
Comment (mgree) [reply src/transform/src/join_implementation.rs:271]: I misunderstood this on our earlier call. These lines go back earlier, though the worst-case scaling factor was added (unconditionally) on the original cardinality PR. #25784 removed the worst-case scaling factor, but not for any particular reason I can see.

Comment (ggevay) [reply src/transform/src/join_implementation.rs:271]: Sorry, I meant the new version of the line, which is not commented out. Could we move it inside the below `if`?

Comment (mgree) [reply src/transform/src/join_implementation.rs:271]: Sure! I don't think it depends on cardinality at all, though.

Comment (ggevay) [reply src/transform/src/join_implementation.rs:271]: Thanks!

**Review by ggevay — APPROVED** [review]

**Pull request merged**